### PR TITLE
Add support for Marlin 2:4 sparsity

### LINKED ISSUE #2098: Sparse Marlin
### Feature request

Sparse-Marlin, an extension of the [Mixed Auto-Regressive Linear (Marlin)](https://github.com/IST-DASLab/marlin/tree/master) dense kernel for 4-bit quantized weights, now with support for 2:4 sparsity.

Specifically, Sparse-Marlin is an extremely optimized GPU kernel performing a matrix multiply and accumulate operation, C=AxB, where A and C are FP16 matrices, and B is a 50% sparse matrix following the 2:4 format, whose values are quantized and stored in INT4 precision. Integrating a 2:4 sparsity scheme on top of the existing 4-bit quantized weight representation offers the opportunity to further enhance the FLOPS/Byte ratio, and push the baseline performance further. Sparse-Marlin boosts the close-to-ideal performance of dense Marlin, and provides near-optimal speedups of around 5.3x, while being more robust to larger batch sizes.

### Motivation

image
image

### Your contribution

https://github.com/IST-DASLab/Sparse-Marlin/blob/main/marlin/marlin_cuda_kernel_nm.cu

Pinging @danieldk since he did the initial marlin PR

I could give it a try if it is wanted, but would also be happy with it if you know

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

This change adds support for 2:4 sparsity when using Marlin quantization. The 2:4 kernel is used when:

* The quantizer is `marlin`;
* the quantizer checkpoint format is `marlin_24`.

Fixes #2098.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/marlin/marlin_kernels/__init__.pyi [hunk 1]
```diff
@@ -19,6 +19,23 @@ def gptq_marlin_gemm(
     """
     ...
 
+def gptq_marlin_24_gemm(
```
Comment (Narsil) [inline server/marlin/marlin_kernels/__init__.pyi:22]: Are we running a fork of marlin ?
Shouldn't we use the classic makefile + commit approach instead (or already made releases if possible) ?

File: server/text_generation_server/layers/marlin.py [hunk 2]
```diff
@@ -215,6 +214,116 @@ def forward(self, A: torch.Tensor) -> torch.Tensor:
         return C
 
 
+GPTQ_MARLIN_24_MIN_THREAD_N = 128
+GPTQ_MARLIN_24_MIN_THREAD_K = 128
+GPTQ_MARLIN_24_MAX_PARALLEL = 64
+GPTQ_MARLIN_24_SUPPORTED_NUM_BITS = [4, 8]
+GPTQ_MARLIN_24_SUPPORTED_GROUP_SIZES = [-1, 128]
```
Comment (Narsil) [inline server/text_generation_server/layers/marlin.py:221]: This is coming from marlin directly I guess ?

File: server/text_generation_server/layers/marlin.py [hunk 3]
```diff
@@ -215,6 +214,116 @@ def forward(self, A: torch.Tensor) -> torch.Tensor:
...
+
+        assert (
+            in_features % GPTQ_MARLIN_24_MIN_THREAD_K == 0
+        ), f"Number of output features ({out_features}) not divisable by {GPTQ_MARLIN_24_MIN_THREAD_K} threads"
+        if groupsize != -1 and in_features % groupsize != 0:
+            raise ValueError(
+                f"Number of input features ({in_features}) not divisable by group size ({groupsize})"
+            )
+
+        self.register_buffer("B", weight.B)
```
Comment (Narsil) [inline server/text_generation_server/layers/marlin.py:290]: NIT.
Any reason to prefer this over `self.B = weight.B`? 

Register buffer changes the default save behavior of pytorch (which we don't care about).
And potentially adds some callbacks (we don't care about either).

**Review by Narsil — APPROVED** [review]
LGTM

File: server/marlin/marlin_kernels/__init__.pyi (hunk 1 above)
Comment (danieldk) [reply server/marlin/marlin_kernels/__init__.pyi:22]: These come from VLLM. We could also directly import them from VLLM if we prefer.

File: server/text_generation_server/layers/marlin.py (hunk 2 above)
Comment (danieldk) [reply server/text_generation_server/layers/marlin.py:221]: Yes, they come from upstream.

File: server/text_generation_server/layers/marlin.py (hunk 3 above)
Comment (danieldk) [reply server/text_generation_server/layers/marlin.py]: Torch module habits, fixed! Also for the other Marlin layers.

**Pull request merged**